storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -96,56 +111,132 @@ func (s *commitLogSource) Read(
 	defer iter.Close()
 
 	var (
-		namespace   = ns.ID()
-		unmerged    = make(map[uint32]map[ts.Hash]encoderMap)
-		bopts       = s.opts.ResultOptions()
-		blopts      = bopts.DatabaseBlockOptions()
-		blockSize   = ns.Options().RetentionOptions().BlockSize()
-		encoderPool = bopts.DatabaseBlockOptions().EncoderPool()
-		errs        = 0
+		namespace    = ns.ID()
+		numShards    = len(shardsTimeRanges)
+		unmerged     = []map[ts.Hash]encodersByTime{}
+		unmergedLock = &sync.RWMutex{}
+		numConc      = s.opts.M3TSZEncodingConcurrency()
+		bopts        = s.opts.ResultOptions()
+		blopts       = bopts.DatabaseBlockOptions()
+		blockSize    = ns.Options().RetentionOptions().BlockSize()
+		encoderPool  = bopts.DatabaseBlockOptions().EncoderPool()
+		workerErrs   = make([]int, numConc)
 	)
+
+	encoderChans := []chan encoderArg{}
+	for i := 0; i < numConc; i++ {
+		encoderChans = append(encoderChans, make(chan encoderArg, encoderChanBufSize))
+	}
+
+	// Spin up numConc background go-routines to handle M3TSZ encoding. This must
+	// happen before we start reading to prevent infinitely blocking writes to
+	// the encoderChans.
+	wg := &sync.WaitGroup{}
+	for workerNum, encoderChan := range encoderChans {
+		wg.Add(1)
+		go s.startM3TSZEncodingWorker(
+			workerNum,
+			encoderChan,
+			&unmerged,
+			unmergedLock,
+			encoderPool,
+			workerErrs,
+			blopts,
+			wg,
+		)
+	}
+
 	for iter.Next() {
 		series, dp, unit, annotation := iter.Current()
-
-		// check if the series belongs to current namespace being bootstrapped
-		if !namespace.Equal(series.Namespace) {
+		if !s.shouldEncodeSeries(namespace, shardsTimeRanges, blockSize, series, dp) {
 			continue
 		}
 
-		ranges, ok := shardsTimeRanges[series.Shard]
-		if !ok {
-			// Not bootstrapping this shard
-			continue
+		// We use the unmerged slice for lookup because its faster than a map.
+		// Shard ID's map directly to indices, but we don't know the highest shard
+		// number upfront so we resize when necessary. We don't need to acquire a
+		// lock when we check the length because this is the only goroutine that
+		// change the size of the slice.
```

Comment:
Technically you could a single range over `shardTimeRanges` to check what the highest shard is up front yeah?